src/main/java/org/logstash/tcp/InputLoop.java

```diff
@@ -0,0 +1,150 @@
+     */
+    public InputLoop(final String host, final int port, final Decoder decoder) {
+        this.decoder = decoder;
+        started = false;
+        try {
+            selector = Selector.open();
+            final ServerSocketChannel accept = ServerSocketChannel.open()
+                .setOption(StandardSocketOptions.SO_REUSEADDR, true)
+                .bind(new InetSocketAddress(host, port));
+            accept.configureBlocking(false);
+            acceptor = accept.register(selector, SelectionKey.OP_ACCEPT);
+        } catch (final IOException ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    @Override
+    public void run() {
+        started = true;
+        try {
+            while (running) {
+                selector.select(100L);
+                final Collection<SelectionKey> keys = selector.selectedKeys();
+                final Iterator<SelectionKey> iter = keys.iterator();
+                while (iter.hasNext()) {
+                    final SelectionKey key = iter.next();
+                    iter.remove();
+                    if (key == acceptor) {
+                        final SocketChannel chan = ((ServerSocketChannel) key.channel()).accept();
+                        if (chan != null) {
+                            chan.configureBlocking(false);
+                            chan.register(
+                                selector, SelectionKey.OP_READ,
+                                ByteBuffer.allocateDirect(BUFFER_SIZE)
+                            );
+                            break;
+                        }
+                    }
+                    if (!read(key) || !key.isValid()) {
+                        cleanup(key);
+                    }
+                }
+            }
+        } catch (final IOException ex) {
+            throw new IllegalStateException(ex);
+        } finally {
+            latch.countDown();
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        running = false;
+        try {
+            if (!started) {
+                latch.countDown();
+            }
+            latch.await();
+            selector.close();
+            acceptor.channel().close();
+        } catch (final InterruptedException ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    /**
+     * Handles Read from Readable SelectionKey.
+     * @param key SelectionKey for which the attached Channel should be read
+     * from
+     * @throws IOException On Failure
+     */
+    private boolean read(final SelectionKey key) {
+        final ByteBuffer buffer = (ByteBuffer) key.attachment();
+        boolean result;
+        try {
+            result = ((ReadableByteChannel) key.channel()).read(buffer) >= 0;
+        } catch (final IOException ignored) {
+            result = false;
+        }
+        buffer.flip();
+        if (buffer.remaining() > 0) {
+            decoder.decode(key);
```

Comment:
@jordansissel and in the same go also, how much data do we want to buffer overall before blocking the I/O thread, given that we're CPU bound here for sure.